dune-precice/dune-precice/couplinginterface.hh

```diff
@@ -0,0 +1,264 @@
+// -*- tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 2 -*-
+// vi: set et ts=4 sw=2 sts=2:
+
+#ifndef DUNE_PRECICE_HH
+#define DUNE_PRECICE_HH
+
+#include <algorithm>
+#include <iostream>
+#include <string>
+#include <vector>
+
+#include <dune/istl/io.hh>
+#include <dune/istl/bvector.hh>
+
+#include <dune/functions/functionspacebases/interpolate.hh>
+
+#include <precice/SolverInterface.hpp>
+
+namespace Dune::preCICE {
+
+  namespace Utilities {
+
+    template <int dim, typename VectorType>
+    static void copy_from_precice_to_dune(VectorType &dune_data,
+                                          const std::vector<double> &precice_data,
+                                          const BlockVector<FieldVector<bool, dim>> is_coupling_boundary)
+    {
+      int iteration_count = 0;
+      for(int i=0; i<dune_data.N(); i++) {
+        for(int j=0; j<dim; j++) {
+          if(is_coupling_boundary[i][j]) { dune_data[i][j] = precice_data[iteration_count]; ++iteration_count; }
+        }
+      }
+    }
+  
+    template <int dim, typename VectorType>
+    static void copy_from_dune_to_precice(const VectorType &dune_data,
+                                          std::vector<double> &precice_data,
+                                          const BlockVector<FieldVector<bool, dim>> is_coupling_boundary)
+    { 
+      int iteration_count = 0;
+      for(int i=0; i<dune_data.N(); i++) {
+        for(int j=0; j<dim; j++) {
+          if(is_coupling_boundary[i][j]) { precice_data[iteration_count] = dune_data[i][j]; ++iteration_count; }
+        }
+      }
+    }
+  } // namespace Utilities  
+
+  template <int dim, typename VectorType, typename ParameterClass>
+  class CouplingInterface
+  {
+
+    private:
+    
+      precice::SolverInterface precice;
+      
+      const BlockVector<FieldVector<bool, dim>> is_coupling_boundary_;
+
+      const std::string mesh_name_;
+      const std::string read_data_name_;
+      const std::string write_data_name_;
+    
+      int mesh_id_;
+      int read_data_id_;
+      int write_data_id_;
+      
+      int n_interface_nodes_;
+      std::vector<double> interface_nodes_positions_;
```

Comment:
I think this should defined locally in the constructor instead of being a member variable because this vector is only used in the constructor, but not afterwards.
```suggestion
```